app/date-select/controller.js

```diff
@@ -0,0 +1,34 @@
+const FormWizardController = require('../../common/controllers/form-wizard')
+const { dateRegex } = require('../../common/helpers/url')
+
+const DateRegExp = new RegExp('/' + dateRegex)
+
+class DateSelectController extends FormWizardController {
+  get(req, res, next) {
+    if (req.query.referrer) {
+      req.sessionModel.reset()
+      req.sessionModel.set('referrer', req.query.referrer)
+      return res.redirect(req.form.options.fullPath)
+    }
+
+    const referrer = req.sessionModel.get('referrer')
+
+    if (!referrer) {
+      return res.redirect('/')
+    }
+
+    res.locals.cancelUrl = referrer
```

Comment:
Maybe these should be set in a [`middlewareLocals`](https://github.com/UKHomeOffice/passports-form-wizard/wiki/HMPO%20Forms%20Flow.pdf) middleware?